Feature - add secret store to IServiceCollection (#216)
arcus-azure/arcus.security: docs/preview/features/secret-store/index.md

                    builder.AddConfiguration(config);
#endif
                    var keyVaultName = config["KeyVault_Name"];

>                     builder.AddEnvironmentVariables()

Reviewer: Let's keep this version maybe?

}
```

> ### Using secret store outside .NET hosting

Reviewer: ```suggestion
### Configuring secret store without .NET host builder
```

> The secret store is also available directly on the `IServiceCollection` for applications that run outside the .NET hosting but still want to use the Arcus secret store.

Reviewer 2: ```suggestion
The secret store is also available directly on the `IServiceCollection` for applications that run without a .NET hosting context but still want to make use of the Arcus secret store.
```

> Just like you would register the secret store on the `HostBuilder`, you can use the `.AddSecretStore` extension to register the secret store:

Reviewer 2: ```suggestion
Just like you would register the secret store on the `HostBuilder`, you can use the `.AddSecretStore` extension method to register the secret store:
```

```csharp
IServiceCollection services = ...

> IConfiguration configuration = ...

Reviewer 2: Just asking, but why not 'completing' the code sample instead of omitting some things ?

Author: Yeah, I added some extra code to make it more clear. Normally, I leave the 'general' types open to not confuse the sample.

// Mark: the secret store uses logging, so make sure that you have set the logging beforehand.

> service.AddLogging(...);

Author: Not sure what to do here. Should we make sure that the secret store can be set up without logging, or should we fail on missing logging infrastructure in the services?

Reviewer 2: My feelings are that you should not impede requirements / dependencies to the consumer of the library.   In other words, although logging is mostly required in projects, usage of the library should not fail if the consumer has not setup logging.

Reviewer 2: I've talked about this with @tomkerkhove  a few days ago regarding logging in libraries.  Logging can be very handy here when troubleshooting, but in normal circumstances, I believe that logs from a library should be as concise as possible.   Libraries should log by default to a Trace / Verbose level imho.

Reviewer: Fully agree on both topics

Reviewer: @stijnmoreels we should change our secret store so that if there is no logger we use a nulllogger

Reviewer: This is maybe also something that should be opted-in/opted-out for in general, but we can discuss that in case that's a good idea

Author: Yes, I also agree on this. I've changed it so logging is only used when it's available.

> services.AddSecretStore(stores =>

Reviewer 2: `AddSecretStore` or `AddSecretProviders` ?

Reviewer: Let's stick with `AddSecretStore` to align with our `ConfigureSecretStore` unless you want to use `ConfigureSecretStore` here as well; but typically services use `Add` as far as I know

Author: Yes, agree, keeping within the .NET naming.

{
    stores.AddEnvironmentVariables();
    
    #if DEBUG
    builder.AddConfiguration(configuration);
    #endif

    var keyVaultName = configuration["KeyVault_Name"];
    stores.AddAzureKeyVaultWithManagedServiceIdentity($"https://{keyVaultName}.vault.azure.net");
});
```

> When the dependency injection container injects the dependent services in the rest of your application, 

Reviewer: This is a bit too much implementation detail imo, what about this:
```suggestion
When your application wants to access a secret, all it has to do is use `ISecretProvider` which will give you access to all the registered secret providers.
```

What do you think?

Author: Yes, no problem!

> the secret store will provide with you an `ISecretProvider` instance that contains the registered secret providers.

Reviewer: ```suggestion
```
